File: sdks/rust/src/stream.rs
pub use crate::types::StreamEvent;
use futures_core::Stream;
use std::pin::Pin;

pub type BoxStream = Pin<Box<dyn Stream<Item = StreamEvent> + Send>>;

/// Collect a streaming response into a single [`ChatResponse`].
///
/// This eliminates the boilerplate of manually matching each
/// [`StreamEventType`] variant and accumulating text, tool calls, and usage
/// tokens.  The returned `ChatResponse` has `model` set to an empty string
/// because the stream events do not carry model information; callers that
/// need the model name should fill it in after the call.
///
/// # Example
///
/// ```ignore
/// let stream = client.stream(messages).await?;
/// let response = motosan_ai::collect_stream(stream).await;
/// println!("{}", response.content);
/// ```
#[cfg(any(
    feature = "anthropic",
    feature = "openai",
    feature = "minimax",
    feature = "ollama_native",
    feature = "gemini",
))]
pub async fn collect_stream(mut stream: BoxStream) -> crate::types::ChatResponse {
    use crate::types::{ChatResponse, StopReason, StreamEventType, ToolCall, Usage};
    use tokio_stream::StreamExt;

    let mut content = String::new();
    let mut tool_calls: Vec<ToolCall> = Vec::new();
    let mut current_tc_id = String::new();
    let mut current_tc_name = String::new();
    let mut current_tc_args = String::new();
    let mut input_tokens: u32 = 0;
    let mut output_tokens: u32 = 0;
    let mut cache_creation_input_tokens: Option<u32> = None;
    let mut cache_read_input_tokens: Option<u32> = None;
    let mut explicit_stop_reason: Option<StopReason> = None;
    let mut session_id: Option<String> = None;
    // Thinking accumulation. `thinking_delta_buf` collects every
    // ThinkingDelta as a fallback in case the provider does not emit
    // ThinkingDone. `thinking_done_buf` holds the explicit final text
    // from the most recent ThinkingDone and takes priority on assembly.
    let mut thinking_delta_buf = String::new();
    let mut thinking_done_buf: Option<String> = None;

    while let Some(event) = stream.next().await {
        if event.session_id.is_some() {
            session_id = event.session_id.clone();
        }
        if event.done {
            // The terminal event may carry a provider-reported stop reason
            // (Anthropic message_delta.stop_reason, OpenAI finish_reason).
            // Capture it before exiting the loop so the heuristic below
            // becomes a fallback only.
            if let Some(reason) = event.stop_reason {
                explicit_stop_reason = Some(reason);
            }
            break;
        }
        match event.event_type {
            StreamEventType::Text => {
                content.push_str(&event.content);
            }
            StreamEventType::Usage => {
                if let Some(ref usage) = event.usage {
                    input_tokens += usage.input_tokens;
                    output_tokens += usage.output_tokens;
                    if let Some(v) = usage.cache_creation_input_tokens {
                        *cache_creation_input_tokens.get_or_insert(0) += v;
                    }
                    if let Some(v) = usage.cache_read_input_tokens {
                        *cache_read_input_tokens.get_or_insert(0) += v;
                    }
                }
            }
            StreamEventType::ToolCallStart => {
                current_tc_id = event.tool_call_id.unwrap_or_default();
                current_tc_name = event.tool_call_name.unwrap_or_default();
                current_tc_args.clear();
            }
            StreamEventType::ToolCallArgs => {
                if let Some(delta) = &event.tool_call_args_delta {
                    current_tc_args.push_str(delta);
                }
            }
            StreamEventType::ToolCallEnd => {
                let input: serde_json::Value = serde_json::from_str(&current_tc_args)
                    .unwrap_or_else(|_| serde_json::json!({}));
                tool_calls.push(ToolCall {
                    id: std::mem::take(&mut current_tc_id),
                    name: std::mem::take(&mut current_tc_name),
                    input,
                });
                current_tc_args.clear();
            }
            StreamEventType::ThinkingDelta => {
                thinking_delta_buf.push_str(&event.content);
            }
            StreamEventType::ThinkingDone => {
                // ThinkingDone carries the full text. Prefer it over the
                // delta accumulator when available — the provider knows
                // the authoritative concatenation. Also clear the delta
                // buffer so a second thinking block starts fresh.
                thinking_done_buf = Some(event.content.clone());
                thinking_delta_buf.clear();
            }
        }
    }

    let stop_reason = match explicit_stop_reason {
        Some(reason) => reason,
        None if tool_calls.is_empty() => StopReason::EndTurn,
        None => StopReason::ToolUse,
    };

    let thinking = match thinking_done_buf {
        Some(text) if !text.is_empty() => Some(text),
        Some(_) => None, // explicit empty thinking block -> treat as none
        None if !thinking_delta_buf.is_empty() => Some(thinking_delta_buf),
        None => None,
    };

    ChatResponse {
        content,
        thinking,
        model: String::new(),
        usage: Usage {
            input_tokens,
            output_tokens,
            cache_creation_input_tokens,
            cache_read_input_tokens,
        },
        stop_reason,
        session_id,
        tool_calls,
    }
}

#[cfg(test)]
#[cfg(any(
    feature = "anthropic",
    feature = "openai",
    feature = "minimax",
    feature = "ollama_native",
    feature = "gemini",
))]
mod thinking_collect_tests {
    use super::*;
    use crate::types::StreamEvent;
    use tokio_stream::iter;

    #[tokio::test]
    async fn collect_stream_accumulates_thinking_into_response_thinking() {
        let events = vec![
            StreamEvent::thinking_delta("Let me "),
            StreamEvent::thinking_delta("think..."),
            StreamEvent::thinking_done("Let me think..."),
            StreamEvent::text("Answer: "),
            StreamEvent::text("42"),
            StreamEvent::done(),
        ];
        let stream: BoxStream = Box::pin(iter(events));
        let resp = collect_stream(stream).await;
        assert_eq!(resp.content, "Answer: 42");
        assert_eq!(
            resp.thinking.as_deref(),
            Some("Let me think..."),
            "thinking field must come from ThinkingDone (or accumulated deltas if no Done)"
        );
    }

    #[tokio::test]
    async fn collect_stream_no_thinking_keeps_thinking_none() {
        let events = vec![StreamEvent::text("hello"), StreamEvent::done()];
        let stream: BoxStream = Box::pin(iter(events));
        let resp = collect_stream(stream).await;
        assert_eq!(resp.content, "hello");
        assert!(resp.thinking.is_none());
    }

    #[tokio::test]
    async fn collect_stream_falls_back_to_accumulated_deltas_if_no_done() {
        // Defensive: if a provider somehow emits ThinkingDelta but skips
        // ThinkingDone, collect_stream still produces a thinking field
        // from the accumulated deltas.
        let events = vec![
            StreamEvent::thinking_delta("A "),
            StreamEvent::thinking_delta("B"),
            StreamEvent::text("ok"),
            StreamEvent::done(),
        ];
        let stream: BoxStream = Box::pin(iter(events));
        let resp = collect_stream(stream).await;
        assert_eq!(resp.thinking.as_deref(), Some("A B"));
        assert_eq!(resp.content, "ok");
    }
}

File: sdks/typescript/src/stream.ts
import type {
  ChatResponse,
  ModelChatResponse,
  ModelStreamDelta,
  ModelToolCall,
  StopReason,
  StreamEvent,
  ToolCall,
  Usage,
} from './types.js'

export type BoxStream = AsyncIterable<StreamEvent>

export function textEvent(content: string): StreamEvent {
  return {
    content,
    done: false,
    eventType: 'text',
  }
}

export function doneEvent(): StreamEvent {
  return {
    content: '',
    done: true,
    eventType: 'text',
  }
}

export function doneWithStopReason(stopReason: StopReason): StreamEvent {
  return {
    content: '',
    done: true,
    eventType: 'text',
    stopReason,
  }
}

export function usageEvent(usage: Usage): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'usage',
    usage,
  }
}

export function toolCallStart(id: string, name: string): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'tool_call_start',
    toolCallId: id,
    toolCallName: name,
  }
}

export function toolCallArgs(delta: string): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'tool_call_args',
    toolCallArgsDelta: delta,
  }
}

export function toolCallArgsWithId(id: string, delta: string): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'tool_call_args',
    toolCallId: id,
    toolCallArgsDelta: delta,
  }
}

export function toolCallEnd(): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'tool_call_end',
  }
}

export function toolCallEndWithId(id: string): StreamEvent {
  return {
    content: '',
    done: false,
    eventType: 'tool_call_end',
    toolCallId: id,
  }
}

export function thinkingDelta(content: string): StreamEvent {
  return {
    content,
    done: false,
    eventType: 'thinking_delta',
  }
}

export function thinkingDone(content: string): StreamEvent {
  return {
    content,
    done: false,
    eventType: 'thinking_done',
  }
}

export async function collectStream(stream: BoxStream): Promise<ChatResponse> {
  let content = ''
  const toolCalls: ToolCall[] = []
  let currentToolId = ''
  let currentToolName = ''
  let currentToolArgs = ''
  let inputTokens = 0
  let outputTokens = 0
  let cacheCreationInputTokens: number | undefined
  let cacheReadInputTokens: number | undefined
  let explicitStopReason: StopReason | undefined
  let thinkingDeltaBuf = ''
  let thinkingDoneBuf: string | undefined

  for await (const event of stream) {
    if (event.done) {
      if (event.stopReason) {
        explicitStopReason = event.stopReason
      }
      break
    }

    switch (event.eventType) {
      case 'text':
        content += event.content
        break

      case 'usage':
        if (event.usage) {
          // Replace-with-fallback, mirroring Python _stream_collect.py:
          // Anthropic message_delta usage is CUMULATIVE, so summing would
          // double-count output tokens. A zero field keeps the previous
          // value (message_delta carries no input_tokens, mapped to 0);
          // an absent cache field keeps the previous cache value.
          inputTokens = event.usage.inputTokens || inputTokens
          outputTokens = event.usage.outputTokens || outputTokens
          if (event.usage.cacheCreationInputTokens !== undefined) {
            cacheCreationInputTokens = event.usage.cacheCreationInputTokens
          }
          if (event.usage.cacheReadInputTokens !== undefined) {
            cacheReadInputTokens = event.usage.cacheReadInputTokens
          }
        }
        break

      case 'tool_call_start':
        currentToolId = event.toolCallId ?? ''
        currentToolName = event.toolCallName ?? ''
        currentToolArgs = ''
        break

      case 'tool_call_args':
        if (event.toolCallArgsDelta) {
          currentToolArgs += event.toolCallArgsDelta
        }
        break

      case 'tool_call_end': {
        let input: unknown = {}
        try {
          input = JSON.parse(currentToolArgs) as unknown
        } catch {
          // Fallback to empty object on parse failure
        }
        toolCalls.push({
          id: event.toolCallId ?? currentToolId,
          name: currentToolName,
          input,
        })
        currentToolArgs = ''
        break
      }

      case 'thinking_delta':
        thinkingDeltaBuf += event.content
        break

      case 'thinking_done':
        thinkingDoneBuf = event.content
        thinkingDeltaBuf = ''
        break
    }
  }

  const stopReason: StopReason = explicitStopReason
    ? explicitStopReason
    : toolCalls.length > 0
      ? 'tool_use'
      : 'end_turn'

  let thinking: string | undefined
  if (thinkingDoneBuf !== undefined) {
    thinking = thinkingDoneBuf.length > 0 ? thinkingDoneBuf : undefined
  } else if (thinkingDeltaBuf.length > 0) {
    thinking = thinkingDeltaBuf
  }

  const usage: Usage = {
    inputTokens,
    outputTokens,
  }
  if (cacheCreationInputTokens !== undefined) {
    usage.cacheCreationInputTokens = cacheCreationInputTokens
  }
  if (cacheReadInputTokens !== undefined) {
    usage.cacheReadInputTokens = cacheReadInputTokens
  }

  return {
    content,
    thinking,
    toolCalls,
    model: '',
    usage,
    stopReason,
  }
}

/** An async iterable of native model stream deltas. Mirrors Rust `BoxModelStream`. */
export type BoxModelStream = AsyncIterable<ModelStreamDelta>

/**
 * Reassemble a native model stream into a ModelChatResponse.
 *
 * Mirrors Rust `collect_model_stream` (stream.rs:155-239). Three rules differ
 * from `collectStream` above and are contract, not implementation detail
 * (milestone D8):
 *
 *   1. `tool_call_done` is AUTHORITATIVE. The accumulated `function_arguments`
 *      / `freeform_input` buffers exist so a caller can render progress; they
 *      are discarded for the completed call and NEVER lowered into it. In
 *      particular a freeform `input` is never JSON-parsed.
 *   2. `usage` REPLACES the running value. (collectStream uses
 *      replace-with-fallback because Anthropic's message_delta usage is
 *      cumulative; the Responses terminal frame is not.)
 *   3. `thinking_done` wins over concatenated `thinking_delta` content, and an
 *      empty `thinking_done` payload means "no thinking", not "empty string".
 *
 * `model` is returned EMPTY — the caller (provider or Client) backfills it
 * from the request/provider default.
 */
export async function collectModelStream(stream: BoxModelStream): Promise<ModelChatResponse> {
  let content = ''
  let thinkingDeltaBuf = ''
  let thinkingDoneBuf: string | undefined
  const functionArguments = new Map<string, string>()
  const freeformInputs = new Map<string, string>()
  const toolCalls: ModelToolCall[] = []
  let usage: Usage = { inputTokens: 0, outputTokens: 0 }
  let explicitStopReason: StopReason | undefined

  for await (const delta of stream) {
    if (delta.type === 'done') {
      explicitStopReason = delta.stopReason
      break
    }

    switch (delta.type) {
      case 'text':
        content += delta.delta
        break

      case 'thinking_delta':
        thinkingDeltaBuf += delta.delta
        break

      case 'thinking_done':
        thinkingDoneBuf = delta.thinking
        thinkingDeltaBuf = ''
        break

      case 'function_arguments':
        functionArguments.set(
          delta.callId,
          (functionArguments.get(delta.callId) ?? '') + delta.delta,
        )
        break

      case 'freeform_input':
        freeformInputs.set(delta.callId, (freeformInputs.get(delta.callId) ?? '') + delta.delta)
        break

      case 'tool_call_done':
        // Discard the bookkeeping buffer; the completed call is the truth.
        if (delta.call.kind === 'function') {
          functionArguments.delete(delta.call.id)
        } else {
          freeformInputs.delete(delta.call.id)
        }
        toolCalls.push(delta.call)
        break

      case 'usage':
        usage = delta.usage
        break
    }
  }

  let thinking: string | undefined
  if (thinkingDoneBuf !== undefined) {
    thinking = thinkingDoneBuf.length > 0 ? thinkingDoneBuf : undefined
  } else if (thinkingDeltaBuf.length > 0) {
    thinking = thinkingDeltaBuf
  }

  const stopReason: StopReason =
    explicitStopReason ?? (toolCalls.length > 0 ? 'tool_use' : 'end_turn')

  const response: ModelChatResponse = {
    content,
    toolCalls,
    model: '',
    usage,
    stopReason,
  }
  if (thinking !== undefined) response.thinking = thinking
  return response
}
